proc_CPMG.py

```diff
@@ -87,43 +68,63 @@ def print_fun(x, f, accepted):
             T=1000.
             )
     zeroorder_rad, firstorder = sol.x
+    angle = (zeroorder_rad)/pi*180 
+    #applying phase shift
     phshift = exp(-1j*2*pi*f_axis*(firstorder*1e-6))
     phshift *= exp(-1j*2*pi*zeroorder_rad)
     s *= phshift
-    print("RELATIVE PHASE SHIFT WAS {:0.1f}\\us and {:0.1f}$^\circ$".format(
-            firstorder,angle(zeroorder_rad)/pi*180))
+    logger.info(strm("RELATIVE PHASE SHIFT WAS {:0.1f}\\us and {:0.1f}$^\circ$".format(firstorder,angle)))
+    fl.show();quit()
     if s['nEchoes',0].data[:].sum().real < 0:
         s *= -1
-    print(ndshape(s))
+    logger.info(strm(ndshape(s)))
     fl.next('after phased - real ft')
     fl.image(s.real)
     fl.next('after phased - imag ft')
     fl.image(s.imag)
-    #data = s['t2':0]
-    data = s['t2':(-200,200)].sum('t2')
-    #data = s['t2':(0,200)].sum('t2')
+
+    #summing along t2 axis and plotting initial decay
+    data = s['t2':(-100,150)].sum('t2')
     fl.next('Echo decay')
-    x = tE_axis
+    x = s.getaxis('nEchoes')
     ydata = data.data.real
     ydata /= max(ydata)
     fl.plot(x,ydata,'-o', alpha=0.7, label='%s'%label_str, human_units=False)
+    print(s.dimlabels)
+    #fl.show();quit()
+    print(ndshape(s))
+    print("Beginning T2 curve")
+    s_sliced = s['nEchoes',(0,60)]['t2',(-50,100)]
+    s = fitdata(s_sliced)
+    M0,T2,tE = sympy.symbols("M_0 T_2 t_E", real=True)
+    s.functional_form = M0*sympy.exp(-tE/T2)
+    print("Functional form", s.functional_form)
+    print("Function string",s.function_string)
+    s.fit_coeff = r_[-1,1,1]
+    fl.next('T2 test')
+    fl.plot(s,'o',label=f.name())
+    print("symbolic variable:",s.symbolic_vars)
```

Comment:
does this work?  if it works w/ the `alec_fitdata` version of `pyspecdata`, then we can leave this alone, since that will soon be merged in ... otherwise and in the future, this is a good example of something that could make an isolated pull request

Reply:
It brings up this error:
```
Traceback (most recent call last):
  File "c:\users\alex guinness\gitstuff\pyspecdata\pyspecdata\core.py", line 6470, in _parse_slices
    axesdict[x] = axesdict[x][y]
IndexError: too many indices for array

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "proc_CPMG.py", line 96, in <module>
    s_sliced = s['nEchoes',(0,60)]['t2',(-50,100)]
  File "c:\users\alex guinness\gitstuff\pyspecdata\pyspecdata\core.py", line 6280, in __getitem__
    slicedict,axesdict,errordict,unitsdict = self._parse_slices(args)
  File "c:\users\alex guinness\gitstuff\pyspecdata\pyspecdata\core.py", line 6472, in _parse_slices
    raise ValueError("axesdict is "+repr(axesdict)+"and I want to set "+repr(x)+" subscript to its "+repr(y)+" value"+ex
plain_error(e))
ValueError: axesdict is {'t2': array([-1984.25388336, -1953.24991643, -1922.24594951, -1891.24198258,
       -1860.23801565, -1829.23404872, -1798.2300818 , -1767.22611487,
       -1736.22214794, -1705.21818101, -1674.21421409, -1643.21024716,
       -1612.20628023, -1581.2023133 , -1550.19834638, -1519.19437945,
       -1488.19041252, -1457.18644559, -1426.18247867, -1395.17851174,
       -1364.17454481, -1333.17057788, -1302.16661096, -1271.16264403,
       -1240.1586771 , -1209.15471017, -1178.15074325, -1147.14677632,
       -1116.14280939, -1085.13884246, -1054.13487554, -1023.13090861,
        -992.12694168,  -961.12297475,  -930.11900783,  -899.1150409 ,
        -868.11107397,  -837.10710704,  -806.10314012,  -775.09917319,
        -744.09520626,  -713.09123933,  -682.08727241,  -651.08330548,
        -620.07933855,  -589.07537162,  -558.0714047 ,  -527.06743777,
        -496.06347084,  -465.05950391,  -434.05553699,  -403.05157006,
        -372.04760313,  -341.0436362 ,  -310.03966928,  -279.03570235,
        -248.03173542,  -217.02776849,  -186.02380157,  -155.01983464,
        -124.01586771,   -93.01190078,   -62.00793386,   -31.00396693,
           0.        ,    31.00396693,    62.00793386,    93.01190078,
         124.01586771,   155.01983464,   186.02380157,   217.02776849,
         248.03173542,   279.03570235,   310.03966928,   341.0436362 ,
         372.04760313,   403.05157006,   434.05553699,   465.05950391,
         496.06347084,   527.06743777,   558.0714047 ,   589.07537162,
         620.07933855,   651.08330548,   682.08727241,   713.09123933,
         744.09520626,   775.09917319,   806.10314012,   837.10710704,
         868.11107397,   899.1150409 ,   930.11900783,   961.12297475,
         992.12694168,  1023.13090861,  1054.13487554,  1085.13884246,
        1116.14280939,  1147.14677632,  1178.15074325,  1209.15471017,
        1240.1586771 ,  1271.16264403,  1302.16661096,  1333.17057788,
        1364.17454481,  1395.17851174,  1426.18247867,  1457.18644559,
        1488.19041252,  1519.19437945,  1550.19834638,  1581.2023133 ,
        1612.20628023,  1643.21024716,  1674.21421409,  1705.21818101,
        1736.22214794,  1767.22611487,  1798.2300818 ,  1829.23404872,
        1860.23801565,  1891.24198258,  1922.24594951,  1953.24991643]), 'nEchoes': array([ 1,  2,  3,  4,  5,  6,  7,
8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64])}and I want to set 'nEchoes' subscript to its (0, 60) value
> Original error (<class 'IndexError'> -- c:\users\alex guinness\gitstuff\pyspecdata\pyspecdata\core.py line 6470):
>       too many indices for array
```
I believe this comes up because in the fitdata_test.py we use proc_IR.py which slices the nddata prior to the t2 curve. in the proc_CPMG we do not. But I am not quite sure what to slice out as I don't want leave out data

Reply:
when you are requesting a pull request to merge into master, it needs to be absolutely functional code.

we are really really suffering here because the branch was left to go for so very long, making so many changes without testing and reviewing them and merging them into master.

this is exacerbating that problem. We can address this in a new pull request.  For now, leave the code alone.

To better explain the problem -- once you get close to resolving my comments, I go through and review the diffs against master, and make comments where I notice that things seem amiss.  Because the diff is **so huge**, it (1) takes me a while to have enough time to do this and (2) I notice things the 2nd or 3rd time through I didn't notice the first time through, so the pull request keeps hanging around like a zombie.

if you think what you have here is really a good start toward using the new fitdata, just `git checkout -b alex_fitdata_attempt` and `git push -u origin alex_fitdata_attempt`, then diff this piece of code back against what you had a few commits ago, and get it back to the point where it works.  Once you've clarified how to use `fitdata`, you can update alex_fitdata_attempt so that it works, and generate a new pull request

Reply:
sounds good